File: notifier.py
import time
from html import escape
from typing import Any, Dict, List, Optional, Set, Tuple

import requests

from config import (
    TELEGRAM_BOT_TOKEN,
    TELEGRAM_CHAT_ID,
    TELEGRAM_LONG_POLL_TIMEOUT,
)
from models import Room, TelegramPollingConflict

TELEGRAM_API_BASE = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}"
TELEGRAM_SEND_MESSAGE_API = f"{TELEGRAM_API_BASE}/sendMessage"
TELEGRAM_GET_UPDATES_API = f"{TELEGRAM_API_BASE}/getUpdates"
TELEGRAM_DELETE_WEBHOOK_API = f"{TELEGRAM_API_BASE}/deleteWebhook"

_TELEGRAM_UPDATES_PREPARED = False

_START_TIME = time.time()
_last_poll_time: Optional[float] = None
_live_streamer_count: int = 0


def update_health_state(live_count: int) -> None:
    """Update health state tracking for /ping responses."""
    global _last_poll_time, _live_streamer_count
    _last_poll_time = time.time()
    _live_streamer_count = live_count


def _handle_ping_command() -> None:
    """Respond to a /ping command with bot health status."""
    import datetime

    uptime_seconds = int(time.time() - _START_TIME)
    hours, remainder = divmod(uptime_seconds, 3600)
    minutes, seconds = divmod(remainder, 60)

    if hours:
        uptime_str = f"{hours}h {minutes}m {seconds}s"
    elif minutes:
        uptime_str = f"{minutes}m {seconds}s"
    else:
        uptime_str = f"{seconds}s"

    if _last_poll_time is not None:
        seconds_ago = int(time.time() - _last_poll_time)
        if seconds_ago < 60:
            last_poll_str = f"{seconds_ago}s ago"
        else:
            last_poll_str = f"{seconds_ago // 60}m ago"
    else:
        last_poll_str = "never"

    text = (
        f"<b>Pong!</b>\n"
        f"Uptime: {uptime_str}\n"
        f"Last poll: {last_poll_str}\n"
        f"Live streamers: {_live_streamer_count}"
    )
    send_telegram(text)


def _process_ping_commands(offset: int) -> int:
    """
    Check for and handle any /ping commands in the Telegram update queue.

    Args:
        offset: Current Telegram update offset

    Returns:
        Updated offset after consuming /ping messages
    """
    current_offset = offset
    try:
        updates, current_offset = get_telegram_updates(
            offset=current_offset, timeout=0
        )
    except TelegramPollingConflict:
        return current_offset

    for update in updates:
        message = update.get('message') or update.get('edited_message') or {}
        chat = message.get('chat') or {}
        if str(chat.get('id', '')) != str(TELEGRAM_CHAT_ID):
            continue

        text = message.get('text', '').strip()
        if text == '/ping':
            _handle_ping_command()

    return current_offset


def send_telegram(text: str) -> bool:
    """Send a message via Telegram bot."""
    try:
        response = requests.post(
            TELEGRAM_SEND_MESSAGE_API,
            json={
                'chat_id': TELEGRAM_CHAT_ID,
                'text': text,
                'parse_mode': 'HTML',
            },
            timeout=10,
        )
        response.raise_for_status()
        return True
    except requests.exceptions.RequestException as error:
        print(f'Warning: Failed to send Telegram message: {error}')
        return False


def get_next_update_offset() -> int:
    """
    Return the next update offset so only future chat messages are consumed.

    Returns:
        Telegram update offset for the next unread message
    """
    updates, next_offset = get_telegram_updates(timeout=0)
    if updates:
        return next_offset
    return 0


def prepare_telegram_updates() -> None:
    """Disable Telegram webhooks once so long polling can use getUpdates."""
    global _TELEGRAM_UPDATES_PREPARED

    if _TELEGRAM_UPDATES_PREPARED:
        return

    try:
        response = requests.post(
            TELEGRAM_DELETE_WEBHOOK_API,
            json={'drop_pending_updates': False},
            timeout=10,
        )
    except requests.exceptions.RequestException as error:
        print(f'Warning: Failed to delete Telegram webhook: {error}')
        return

    description = _extract_telegram_description(response)
    if response.status_code == 409:
        raise TelegramPollingConflict(_build_polling_conflict_message(description))

    if not response.ok:
        print(
            'Warning: Telegram deleteWebhook failed: '
            f'{response.status_code} {description}'
        )
        return

    try:
        payload = response.json()
    except ValueError:
        payload = {'ok': True}

    if not payload.get('ok', False):
        print(f'Warning: Telegram deleteWebhook returned an error: {payload}')
        return

    _TELEGRAM_UPDATES_PREPARED = True


def get_telegram_updates(
    offset: Optional[int] = None,
    timeout: int = TELEGRAM_LONG_POLL_TIMEOUT,
) -> Tuple[List[Dict[str, Any]], int]:
    """
    Fetch Telegram bot updates.

    Args:
        offset: Optional Telegram update offset
        timeout: Long-poll timeout in seconds

    Returns:
        Tuple of updates and the next update offset
    """
    prepare_telegram_updates()
    params: Dict[str, Any] = {'timeout': timeout}
    if offset is not None:
        params['offset'] = offset

    try:
        response = requests.get(
            TELEGRAM_GET_UPDATES_API,
            params=params,
            timeout=timeout + 10,
        )
    except requests.exceptions.RequestException as error:
        print(f'Warning: Failed to fetch Telegram updates: {error}')
        return [], offset or 0

    description = _extract_telegram_description(response)
    if response.status_code == 409:
        raise TelegramPollingConflict(_build_polling_conflict_message(description))

    if not response.ok:
        print(
            'Warning: Telegram updates request failed: '
            f'{response.status_code} {description}'
        )
        return [], offset or 0

    payload = response.json()

    if not payload.get('ok'):
        print(f'Warning: Telegram API returned an error: {payload}')
        return [], offset or 0

    updates = payload.get('result', [])
    next_offset = offset or 0
    if updates:
        next_offset = int(updates[-1]['update_id']) + 1

    return updates, next_offset


def wait_for_chat_message(offset: int) -> Tuple[Optional[str], int]:
    """
    Wait for the next text message from the configured Telegram chat.

    Args:
        offset: Telegram update offset to start from

    Returns:
        Tuple of message text and next update offset
    """
    current_offset = offset

    while True:
        updates, current_offset = get_telegram_updates(offset=current_offset)
        if not updates:
            time.sleep(5)
            continue

        for update in updates:
            message = update.get('message') or update.get('edited_message') or {}
            chat = message.get('chat') or {}
            if str(chat.get('id', '')) != str(TELEGRAM_CHAT_ID):
                continue

            text = message.get('text', '').strip()
            if text:
                return text, current_offset


def _extract_telegram_description(response: requests.Response) -> str:
    """Extract Telegram API error descriptions safely."""
    try:
        payload = response.json()
    except ValueError:
        return response.text.strip()

    return str(payload.get('description', '')).strip()


def _build_polling_conflict_message(description: str) -> str:
    """Return a readable explanation for Telegram getUpdates conflicts."""
    if 'webhook' in description.lower():
        return (
            'Telegram bot polling is blocked by an active webhook. '
            'Disable the webhook for this bot token and try again.'
        )

    return (
        'Telegram getUpdates is already being consumed by another process for '
        'this bot token. Stop the other bot instance or use a different bot token.'
    )


def notify_new_live(rooms: List[Room], previous_live: Optional[Set[str]]) -> Set[str]:
    """
    Send Telegram notifications for newly live streamers.

    Args:
        rooms: List of Room objects
        previous_live: Set of room_ids that were live in previous check

    Returns:
        Set of currently live room IDs
    """
    current_live = {r.room_id for r in rooms if r.is_live}

    # Only notify on new streams (skip first run to avoid spam)
    if previous_live is not None:
        new_live_ids = current_live - previous_live
        for room in rooms:
            if room.room_id in new_live_ids:
                text = (
                    f'<b>{escape(room.streamer_name)}</b> is now live!\n'
                    f'{escape(room.room_name)}\n'
                    f'Category: {escape(room.area_name)}\n'
                    f"<a href=\"{room.url}\">Watch</a>"
                )
                send_telegram(text)
                print(f"  Notified: {room.streamer_name} is live")

    return current_live

File: notifier.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	telegramHTTPClient httpDoer = &http.Client{}

	telegramPreparationMu sync.Mutex
	telegramUpdatesReady  bool

	healthMu          sync.RWMutex
	startTime         = time.Now()
	lastPollTime      time.Time
	hasLastPoll       bool
	liveStreamerCount int

	nowFunc          = time.Now
	sleepContextFunc = sleepContext
	sendTelegramFunc = sendTelegram
)

type telegramChat struct {
	ID json.RawMessage `json:"id"`
}

type telegramMessage struct {
	Chat telegramChat `json:"chat"`
	Text string       `json:"text"`
}

type telegramUpdate struct {
	UpdateID      int64            `json:"update_id"`
	Message       *telegramMessage `json:"message"`
	EditedMessage *telegramMessage `json:"edited_message"`
}

type telegramAPIResponse struct {
	OK          bool             `json:"ok"`
	Description json.RawMessage  `json:"description"`
	Result      []telegramUpdate `json:"result"`
}

func telegramAPIEndpoint(method string) string {
	return "https://api.telegram.org/bot" + telegramBotToken + "/" + method
}

// updateHealthState records information returned by the /ping command.
func updateHealthState(liveCount int) {
	healthMu.Lock()
	defer healthMu.Unlock()
	lastPollTime = nowFunc()
	hasLastPoll = true
	liveStreamerCount = liveCount
}

func handlePingCommand(ctx context.Context) {
	now := nowFunc()
	healthMu.RLock()
	started := startTime
	polledAt := lastPollTime
	wasPolled := hasLastPoll
	liveCount := liveStreamerCount
	healthMu.RUnlock()

	uptimeSeconds := int(now.Sub(started).Seconds())
	hours := uptimeSeconds / 3600
	minutes := (uptimeSeconds % 3600) / 60
	seconds := uptimeSeconds % 60

	var uptime string
	switch {
	case hours != 0:
		uptime = fmt.Sprintf("%dh %dm %ds", hours, minutes, seconds)
	case minutes != 0:
		uptime = fmt.Sprintf("%dm %ds", minutes, seconds)
	default:
		uptime = fmt.Sprintf("%ds", seconds)
	}

	lastPoll := "never"
	if wasPolled {
		secondsAgo := int(now.Sub(polledAt).Seconds())
		if secondsAgo < 60 {
			lastPoll = fmt.Sprintf("%ds ago", secondsAgo)
		} else {
			lastPoll = fmt.Sprintf("%dm ago", secondsAgo/60)
		}
	}

	text := fmt.Sprintf(
		"<b>Pong!</b>\nUptime: %s\nLast poll: %s\nLive streamers: %d",
		uptime,
		lastPoll,
		liveCount,
	)
	sendTelegramFunc(ctx, text)
}

func processPingCommands(ctx context.Context, offset int64, timeout int) (int64, error) {
	updates, nextOffset, err := getTelegramUpdates(ctx, &offset, timeout)
	if err != nil {
		var conflict *TelegramPollingConflict
		if errors.As(err, &conflict) {
			return offset, nil
		}
		return offset, err
	}

	for _, update := range updates {
		message := update.Message
		if message == nil {
			message = update.EditedMessage
		}
		if message == nil || telegramChatIDString(message.Chat.ID) != telegramChatID {
			continue
		}
		if strings.TrimSpace(message.Text) == "/ping" {
			handlePingCommand(ctx)
		}
	}
	return nextOffset, nil
}

// sendTelegram sends one HTML-formatted message to the configured chat.
func sendTelegram(ctx context.Context, text string) bool {
	body, _ := json.Marshal(map[string]string{
		"chat_id":    telegramChatID,
		"text":       text,
		"parse_mode": "HTML",
	})

	requestContext, cancel := context.WithTimeout(ctx, telegramRequestGracePeriod)
	defer cancel()
	req, err := http.NewRequestWithContext(
		requestContext,
		http.MethodPost,
		telegramAPIEndpoint("sendMessage"),
		bytes.NewReader(body),
	)
	if err == nil {
		req.Header.Set("Content-Type", "application/json")
		var response *http.Response
		response, err = telegramHTTPClient.Do(req)
		if err == nil {
			defer response.Body.Close()
			_, _ = io.Copy(io.Discard, response.Body)
			if response.StatusCode >= http.StatusOK && response.StatusCode < http.StatusBadRequest {
				return true
			}
			err = fmt.Errorf("HTTP status %s", response.Status)
		}
	}

	if ctx.Err() == nil {
		fmt.Printf("Warning: Failed to send Telegram message: %v\n", err)
	}
	return false
}

func getNextUpdateOffset(ctx context.Context) (int64, error) {
	updates, nextOffset, err := getTelegramUpdates(ctx, nil, 0)
	if err != nil {
		return 0, err
	}
	if len(updates) > 0 {
		return nextOffset, nil
	}
	return 0, nil
}

// prepareTelegramUpdates disables webhooks once so getUpdates can long poll.
func prepareTelegramUpdates(ctx context.Context) error {
	telegramPreparationMu.Lock()
	defer telegramPreparationMu.Unlock()
	if telegramUpdatesReady {
		return nil
	}

	body := []byte(`{"drop_pending_updates":false}`)
	requestContext, cancel := context.WithTimeout(ctx, telegramRequestGracePeriod)
	defer cancel()
	req, err := http.NewRequestWithContext(
		requestContext,
		http.MethodPost,
		telegramAPIEndpoint("deleteWebhook"),
		bytes.NewReader(body),
	)
	if err != nil {
		fmt.Printf("Warning: Failed to delete Telegram webhook: %v\n", err)
		return nil
	}
	req.Header.Set("Content-Type", "application/json")

	response, err := telegramHTTPClient.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		fmt.Printf("Warning: Failed to delete Telegram webhook: %v\n", err)
		return nil
	}
	defer response.Body.Close()
	responseBody, readErr := io.ReadAll(response.Body)
	if readErr != nil {
		fmt.Printf("Warning: Failed to delete Telegram webhook: %v\n", readErr)
		return nil
	}

	description := extractTelegramDescription(responseBody)
	if response.StatusCode == http.StatusConflict {
		return &TelegramPollingConflict{Message: buildPollingConflictMessage(description)}
	}
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusBadRequest {
		fmt.Printf(
			"Warning: Telegram deleteWebhook failed: %d %s\n",
			response.StatusCode,
			description,
		)
		return nil
	}

	var payload map[string]any
	if err := json.Unmarshal(responseBody, &payload); err != nil {
		telegramUpdatesReady = true
		return nil
	}
	if ok, _ := payload["ok"].(bool); !ok {
		fmt.Printf("Warning: Telegram deleteWebhook returned an error: %s\n", strings.TrimSpace(string(responseBody)))
		return nil
	}

	telegramUpdatesReady = true
	return nil
}

func getTelegramUpdates(
	ctx context.Context,
	offset *int64,
	timeout int,
) ([]telegramUpdate, int64, error) {
	if err := prepareTelegramUpdates(ctx); err != nil {
		return nil, offsetValue(offset), err
	}

	endpoint, err := url.Parse(telegramAPIEndpoint("getUpdates"))
	if err != nil {
		return nil, offsetValue(offset), err
	}
	query := endpoint.Query()
	query.Set("timeout", strconv.Itoa(timeout))
	if offset != nil {
		query.Set("offset", strconv.FormatInt(*offset, 10))
	}
	endpoint.RawQuery = query.Encode()

	requestTimeout := time.Duration(timeout)*time.Second + telegramRequestGracePeriod
	requestContext, cancel := context.WithTimeout(ctx, requestTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(requestContext, http.MethodGet, endpoint.String(), nil)
	if err != nil {
		return nil, offsetValue(offset), err
	}

	response, err := telegramHTTPClient.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return nil, offsetValue(offset), ctx.Err()
		}
		fmt.Printf("Warning: Failed to fetch Telegram updates: %v\n", err)
		return nil, offsetValue(offset), nil
	}
	defer response.Body.Close()
	responseBody, readErr := io.ReadAll(response.Body)
	if readErr != nil {
		fmt.Printf("Warning: Failed to fetch Telegram updates: %v\n", readErr)
		return nil, offsetValue(offset), nil
	}

	description := extractTelegramDescription(responseBody)
	if response.StatusCode == http.StatusConflict {
		return nil, offsetValue(offset), &TelegramPollingConflict{
			Message: buildPollingConflictMessage(description),
		}
	}
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusBadRequest {
		fmt.Printf(
			"Warning: Telegram updates request failed: %d %s\n",
			response.StatusCode,
			description,
		)
		return nil, offsetValue(offset), nil
	}

	var payload telegramAPIResponse
	if err := json.Unmarshal(responseBody, &payload); err != nil {
		return nil, offsetValue(offset), fmt.Errorf("invalid Telegram response: %w", err)
	}
	if !payload.OK {
		fmt.Printf("Warning: Telegram API returned an error: %s\n", strings.TrimSpace(string(responseBody)))
		return nil, offsetValue(offset), nil
	}

	nextOffset := offsetValue(offset)
	if len(payload.Result) > 0 {
		nextOffset = payload.Result[len(payload.Result)-1].UpdateID + 1
	}
	return payload.Result, nextOffset, nil
}

func waitForChatMessage(ctx context.Context, offset int64) (string, int64, error) {
	currentOffset := offset
	for {
		updates, nextOffset, err := getTelegramUpdates(ctx, &currentOffset, telegramLongPollTimeout)
		if err != nil {
			return "", currentOffset, err
		}
		currentOffset = nextOffset
		if len(updates) == 0 {
			if err := sleepContextFunc(ctx, 5*time.Second); err != nil {
				return "", currentOffset, err
			}
			continue
		}

		for _, update := range updates {
			message := update.Message
			if message == nil {
				message = update.EditedMessage
			}
			if message == nil || telegramChatIDString(message.Chat.ID) != telegramChatID {
				continue
			}
			if text := strings.TrimSpace(message.Text); text != "" {
				return text, currentOffset, nil
			}
		}
	}
}

func extractTelegramDescription(body []byte) string {
	var payload map[string]json.RawMessage
	if err := json.Unmarshal(body, &payload); err != nil {
		return strings.TrimSpace(string(body))
	}
	description, ok := payload["description"]
	if !ok {
		return ""
	}
	return strings.TrimSpace(rawScalarString(description, ""))
}

func buildPollingConflictMessage(description string) string {
	if strings.Contains(strings.ToLower(description), "webhook") {
		return "Telegram bot polling is blocked by an active webhook. " +
			"Disable the webhook for this bot token and try again."
	}
	return "Telegram getUpdates is already being consumed by another process for " +
		"this bot token. Stop the other bot instance or use a different bot token."
}

// notifyNewLive sends one message for each room that became live.
func notifyNewLive(ctx context.Context, rooms []Room, previousLive map[string]struct{}) map[string]struct{} {
	currentLive := currentLiveRoomIDs(rooms)
	if previousLive != nil {
		for _, room := range rooms {
			if _, liveNow := currentLive[room.RoomID]; !liveNow {
				continue
			}
			if _, wasLive := previousLive[room.RoomID]; wasLive {
				continue
			}

			text := fmt.Sprintf(
				"<b>%s</b> is now live!\n%s\nCategory: %s\n<a href=\"%s\">Watch</a>",
				htmlEscape(room.StreamerName),
				htmlEscape(room.RoomName),
				htmlEscape(room.AreaName),
				room.URL,
			)
			sendTelegramFunc(ctx, text)
			fmt.Printf("  Notified: %s is live\n", room.StreamerName)
		}
	}
	return currentLive
}

// notifyStreamEnd sends one message for each room that stopped streaming.
func notifyStreamEnd(ctx context.Context, rooms []Room, previousLive map[string]struct{}) map[string]struct{} {
	currentLive := currentLiveRoomIDs(rooms)
	if previousLive == nil {
		return currentLive
	}

	roomLookup := make(map[string]Room, len(rooms))
	for _, room := range rooms {
		roomLookup[room.RoomID] = room
	}
	for roomID := range previousLive {
		if _, stillLive := currentLive[roomID]; stillLive {
			continue
		}

		streamerName := roomID
		if room, ok := roomLookup[roomID]; ok {
			streamerName = htmlEscape(room.StreamerName)
		}
		sendTelegramFunc(ctx, fmt.Sprintf("<b>%s</b> has ended their stream.", streamerName))
		fmt.Printf("  Notified: %s ended stream\n", streamerName)
	}
	return currentLive
}

func currentLiveRoomIDs(rooms []Room) map[string]struct{} {
	currentLive := map[string]struct{}{}
	for _, room := range rooms {
		if room.IsLive {
			currentLive[room.RoomID] = struct{}{}
		}
	}
	return currentLive
}

func htmlEscape(text string) string {
	text = strings.ReplaceAll(text, "&", "&amp;")
	text = strings.ReplaceAll(text, "<", "&lt;")
	text = strings.ReplaceAll(text, ">", "&gt;")
	text = strings.ReplaceAll(text, "\"", "&quot;")
	return strings.ReplaceAll(text, "'", "&#x27;")
}

func telegramChatIDString(raw json.RawMessage) string {
	return rawScalarString(raw, "")
}

func offsetValue(offset *int64) int64 {
	if offset == nil {
		return 0
	}
	return *offset
}

func sleepContext(ctx context.Context, delay time.Duration) error {
	timer := time.NewTimer(delay)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}
